src/libmongoc/src/mongoc/mongoc-openssl.c

```diff
@@ -844,8 +874,12 @@ _mongoc_ocsp_tlsext_status (SSL *ssl, mongoc_openssl_ocsp_opt_t *opts)
     * 2. Next, the function verifies the signature of the basic response.
     * 3. Finally, the function validates the signer cert, constructing the
     * validation path via the untrusted cert chain.
+    *
+    * cert_chain has already been verified. Use OCSP_TRUSTOTHER so the signer
+    * certificate can be considered verified if it is in cert_chain.
     */
-   if (OCSP_basic_verify (basic, cert_chain, store, 0) != OCSP_VERIFY_SUCCESS) {
+   if (OCSP_basic_verify (basic, cert_chain, store, OCSP_TRUSTOTHER) !=
+       OCSP_VERIFY_SUCCESS) {
```

Comment:
After the `Host` header was included, I had noticed that evergreen hosts were still failing to validate the response. It was due to the difference of installed trusted certificates.

The kms.us-east-1.amazonaws.com returns a chain of four certs: 1 -> 2 -> 3 -> 4.
1 is the peer certificate, and 2 is the issuer.
The OCSP response is signed by 2.
Locally, I have 3 (Amazon Root CA) installed.
On Evergreen, 3 is not installed, but 4 is.

`OCSP_basic_verify` attempts to build the chain from the OCSP response's signing certificate (2) to verify the signature. Since I locally have 3 installed, it can fetch that directly from `store`. But on Evergreen, `store` does not have 3. But 3 is included in `cert_chain`, since that comes from the chain that the server presents.

`OCSP_basic_verify` considers `cert_chain` an untrusted list of certificates by default, and will not use it to verify the OCSP response. However, we've already verified `cert_chain`. We can use OCSP_TRUSTOTHER to indicate `cert_chain` contains verified certificates, and can be used to verify the OCSP signature certificate.

See also: https://www.mail-archive.com/openssl-users@openssl.org/msg22791.html for more documentation of the flags.

Reply:
>  However, we've already verified `cert_chain`.

Where did we verify `cert_chain`?

Reply:
This chain is obtained from `_get_verified_chain`. Since the OCSP check happens after `BIO_do_handshake`, at that the server certificate chain has already been verified.

Reply:
Ah, I see.